Documentation of first section

Fizzixnerd opened the pull request:
Added LDoc documentation of first section of tiles.lua.

Added generated documentation in doc/ directory.

Fizzixnerd committed:
Added documentation for the first section.

Added author, copyright, license links.

Documented getStringWidth, getMaxStringLength, checkProperties[...].

Documented locals findBounds, ALIGNMENT, _checkProperty, calculateProperty.

TODO: Document locals formatProperty{In,Out}.

Fizzixnerd committed:
Added generated docs generated by LDoc.

blunty666 commented on tiles.lua:
```diff
@@ -273,10 +331,15 @@ local calculateProperty = {
 local drawObjectWithType = {}
 local objectMetatables = {}
 
+-- Metaprogramming magic below.
```
This is where I'm not sure LDoc is the right approach for this. Because of the way the objects are generated when the API first loads, there is no explicit reference in the code to the individual functions they have. We could just end up making the API file larger without adding any real value...

Do you know a way around this using LDoc?

Fizzixnerd commented:
Yes, I will submit another pull request with the work so far on the metaprogramming bit.  It's not complete, but should illustrate that it is possible to do.  When the references are completed between the Tiles and the particular Objects, I think it will add considerable value.

Fizzixnerd committed:
Nearly finished commenting the base Object.

config.ld: added small and basic configuration file for LDoc.

doc/index.html: updated generated documentation.

tiles.lua: documented most of the base Object interface with LDoc
markup.

TODO: Finish markup of base Objects.

Fizzixnerd commented:
I would, by the way, recommend taking a look at the generated doc/index.html file too, to see what the generated docs actually look like.  You could probably stick them on github.io or something, if you wanted.

blunty666 commented:
Ok, let me know how you get on. I've had a look at the generated html and it looks really good, I like the format so far. I also like the fact you can tell it to ignore documentation on local functions / variables, so you can have two versions, one that only details stuff you can access from the API, and one that covers all the internals as well.

blunty666 commented on tiles.lua:
```diff
@@ -143,11 +167,41 @@ local _checkProperty = {
 end
 
+--- No idea.
+-- @todo What exactly do these do...?
```
These translate the arguments the user passes into the API into arguments the API is expecting.

For example, with alignment it gets passed as object:SetObjectAlignment(horizontal, vertical), but the API stores it as a table {horizontal, vertical}, so this function does the neccessary conversion.

For the points, it's simply to strip away any additional indexes from the tables passed in by the user. It also means a completely new copy is made just in case the user modifies the table after passing it into the API.

The formatPropertyOut functions do the same, just in reverse.

Fizzixnerd committed:
Updated docs to reflect work done.

Finished base Object docs.

Finished Box docs.

Finished Fluid docs.

Finished GradientBox docs.

Finished GradientLine docs.

Started GradientLineList docs (unfinished).

Fizzixnerd commented:
Alright.  I added some mention of the copying to the docs.

edit: API -> docs
